odoo/orm/fields_properties.py

```diff
@@ -669,41 +671,35 @@ def convert_to_record(self, value, record):
-
-            # check if the model still exists in the environment, the module of the
-            # model might have been uninstalled so the model might not exist anymore
-            property_model = property_definition.get('comodel')
-            if property_model and property_model not in record.env:
-                property_definition['comodel'] = property_model = False
-
-            if not property_model and 'domain' in property_definition:
-                del property_definition['domain']
-
-            if property_definition.get('type') in ('selection', 'tags'):
+            property_definition = copy.deepcopy(property_definition)
+
+            type_ = property_definition.get('type')
+            
+            if type_ in ('many2one', 'many2many'):
+                # check if the model still exists in the environment, the module of the
+                # model might have been uninstalled so the model might not exist anymore
+                property_model = property_definition.get('comodel')
+                if property_model not in record.env:
+                    property_definition['comodel'] = False
+                    property_definition.pop('domain', None)
+                elif property_domain := property_definition.get('domain'):
+                    # some fields in the domain might have been removed
+                    # (e.g. if the module has been uninstalled)
+                    # check if the domain is still valid
+                    try:
+                        expression.expression(
+                            ast.literal_eval(property_domain),
+                            record.env[property_model],
+                        )
+                    except ValueError:
+                        del property_definition['domain']
+
+            elif type_ in ('selection', 'tags'):
                 # always set at least an empty array if there's no option
-                key = property_definition['type']
-                property_definition[key] = property_definition.get(key) or []
-
-            property_domain = property_definition.get('domain')
-            if property_domain:
-                # some fields in the domain might have been removed
-                # (e.g. if the module has been uninstalled)
-                # check if the domain is still valid
-                try:
-                    expression.expression(
-                        ast.literal_eval(property_domain),
-                        record.env[property_model],
-                    )
-                except ValueError:
-                    del property_definition['domain']
+                property_definition[type_] = property_definition.get(type_) or []
 
             result.append(property_definition)
 
-            for property_parameter, allowed_types in self.PROPERTY_PARAMETERS_MAP.items():
-                if property_definition.get('type') not in allowed_types:
-                    property_definition.pop(property_parameter, None)
```

Comment:
The logic is handled in `_validate_properties_definition`, so for new data after this commit. The property definition should be clean.
but for the already contaminated data, we should either 
1. keep the code when read
or 2. add an upgrade script to clean

Reply:
better to do the option 2 I think